Add Qiskit native QPY ParameterExpression serialization

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

With the release of symengine 0.13.0 we discovered a version dependence on the payload format used for serializing symengine expressions. This was worked around in #13251 but this is not a sustainable solution and only works for symengine 0.11.0 and 0.13.0 (there was no 0.12.0). While there was always the option to use sympy to serialize the underlying symbolic expression (there is a `use_symengine` flag on `qpy.dumps` you can set to `False` to do this) the sympy serialzation has several tradeoffs most importantly is much higher runtime overhead. To solve the issue moving forward a qiskit native representation of the parameter expression object is necessary for serialization.

This commit bumps the QPY format version to 13 and adds a new serialization format for ParameterExpression objects. This new format is a serialization of the API calls made to ParameterExpression that resulted in the creation of the underlying object. To facilitate this the ParameterExpression class is expanded to store an internal "replay" record of the API calls used to construct the ParameterExpression object. This internal list is what gets serialized by QPY and then on deserialization the "replay" is replayed to reconstruct the expression object. This is a different approach to the previous QPY representations of the ParameterExpression objects which instead represented the internal state stored in the ParameterExpression object with the symbolic expression from symengine (or a sympy copy of the expression). Doing this directly in Qiskit isn't viable though because symengine's internal expression tree is not exposed to Python directly. There isn't any method (private or public) to walk the expression tree to construct a serialization format based off of it. Converting symengine to a sympy expression and then using sympy's API to walk the expression tree is a possibility but that would tie us to sympy which would be problematic for #13267 and #13131, have significant runtime overhead, and it would be just easier to rely on sympy's native serialization tools.

The tradeoff with this approach is that it does increase the memory overhead of the `ParameterExpression` class because for each element in the expression we have to store a record of it. Depending on the depth of the expression tree this also could be a lot larger than symengine's internal representation as we store the raw api calls made to create the ParameterExpression but symengine is likely simplifying it's internal representation as it builds it out. But I personally think this tradeoff is worthwhile as it ties the serialization format to the Qiskit objects and APIs instead of relying on a 3rd party library. This also gives us the flexibility of changing the internal symbolic expression library internally in the future if we decide to stop using symengine at any point.

### Details and comments

Fixes #13252

TODO:

- [x] Add QPY v13 format documentation
- [x] Add release notes
- [x] Expand test coverage. We need to ensure we have thorough coverage of all supported expression operations which we weren't previously exercising through QPY.

mtreinish committed:
Add Qiskit native QPY ParameterExpression serialization

With the release of symengine 0.13.0 we discovered a version dependence
on the payload format used for serializing symengine expressions. This
was worked around in #13251 but this is not a sustainable solution and
only works for symengine 0.11.0 and 0.13.0 (there was no 0.12.0). While
there was always the option to use sympy to serialize the underlying
symbolic expression (there is a `use_symengine` flag on `qpy.dumps` you
can set to `False` to do this) the sympy serialzation has several
tradeoffs most importantly is much higher runtime overhead. To solve
the issue moving forward a qiskit native representation of the parameter
expression object is necessary for serialization.

This commit bumps the QPY format version to 13 and adds a new
serialization format for ParameterExpression objects. This new format
is a serialization of the API calls made to ParameterExpression that
resulted in the creation of the underlying object. To facilitate this
the ParameterExpression class is expanded to store an internal "replay"
record of the API calls used to construct the ParameterExpression
object. This internal list is what gets serialized by QPY and then on
deserialization the "replay" is replayed to reconstruct the expression
object. This is a different approach to the previous QPY representations
of the ParameterExpression objects which instead represented the internal
state stored in the ParameterExpression object with the symbolic
expression from symengine (or a sympy copy of the expression). Doing
this directly in Qiskit isn't viable though because symengine's internal
expression tree is not exposed to Python directly. There isn't any
method (private or public) to walk the expression tree to construct
a serialization format based off of it. Converting symengine to a sympy
expression and then using sympy's API to walk the expression tree is a
possibility but that would tie us to sympy which would be problematic
for #13267 and #13131, have significant runtime overhead, and it would
be just easier to rely on sympy's native serialization tools.

The tradeoff with this approach is that it does increase the memory
overhead of the `ParameterExpression` class because for each element
in the expression we have to store a record of it. Depending on the
depth of the expression tree this also could be a lot larger than
symengine's internal representation as we store the raw api calls made
to create the ParameterExpression but symengine is likely simplifying
it's internal representation as it builds it out. But I personally think
this tradeoff is worthwhile as it ties the serialization format to the
Qiskit objects instead of relying on a 3rd party library. This also
gives us the flexibility of changing the internal symbolic expression
library internally in the future if we decide to stop using symengine
at any point.

Fixes #13252

mtreinish commented on qiskit/qpy/formats.py:
```diff
@@ -259,6 +259,18 @@
+# PARAM_EXPR_V2 = namedtuple("PARAM_EXPR_V2", ["size"])
+# PARAM_EXPR_V2_PACK = "!H"
+# PARAM_EXPR_V2_SIZE = struct.calcsize(PARAM_EXPR_V2_PACK)
+
```
```suggestion
```

mtreinish committed:
Remove stray comment

mtreinish committed:
Add format documentation

mtreinish committed:
Add release note

mtreinish committed:
Merge branch 'main' into qpy-paramexpr-native

mtreinish committed:
Add test and fix some issues with recursive expressions

mtreinish committed:
Add int type for operands

mtreinish committed:
Merge remote-tracking branch 'origin/main' into qpy-paramexpr-native

mtreinish committed:
Add dedicated subs test

mtreinish commented:
The state of this is it's stuck on the overly complex expression example test I added in: https://github.com/Qiskit/qiskit/pull/13356/commits/2b352d705ab30f2106f5668ec07a3fbe0bbfcd41 I'm thinking it might be worth approaching the deserialization in a slightly different approach instead of doing nested expressions recursively like it currently is. If anyone has bandwidth and can come up with a fix before me for that test, feel free to just push to the branch. I think when that test is passing we probably have enough coverage between that new test and the expression tests that already existed for circuits and pulse schedules (we might want to convert those expression tests to circuit equivalents as part of the pulse removal in 2.0).

mtreinish committed:
Pivot to stack based postfix/rpn deserialization

This commit changes how the deserialization works to use a postfix
stack based approach. Operands are push on the stack and then popped off
based on the operation being run. The result of the operation is then
pushed on the stack. This handles nested objects much more cleanly than
the recursion based approach because we just keep pushing on the stack
instead of recursing, making the accounting much simpler. After the
expression payload is finished being processed there will be a single
value on the stack and that is returned as the final expression.

mtreinish committed:
Merge branch 'main' into qpy-paramexpr-native

mtreinish commented:
Ok, I fixed the issue from last week with: https://github.com/Qiskit/qiskit/pull/13356/commits/a903387f435294cc5a85a2e39d119a83e4864578 which switches to a stack based approach instead of a recursive one. I think this should be ready now.

ElePT commented on qiskit/circuit/parameterexpression.py:
```diff
@@ -30,12 +33,80 @@
+    EXP = 10
+    LOG = 11
+    SIGN = 12
+    DERIV = 13
```
Minor question, any reason why you chose `DERIV` instead of `GRAD`? Is it because it was closer to the symengine notation?

mtreinish replied:
Yeah it was more or less to match the functionality to what symengine called it. I can switch it to `GRAD` or `GRADIENT` if you'd prefer, the only thing that really matters here is the `13`.

ElePT replied:
this is 100% nitpick but for some reason I would prefer `GRAD`, yes, but it's not a super strong request

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13356/commits/8ae658932de61ad5c06e70572bb9c8e0d23a5904

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -322,6 +342,139 @@
+The first change is the :ref:`qpy_param_expr_v3` payload for the ``expr_size`` bytes following
+the header is now comprised of an array of ``PARAM_EXPR_ELEM_V13`` structs. The intent is for the
+array of elements to be read one struct at a time that describe the calls to make to reconstruct
+the :class:`.ParameterExpression`.
```
```suggestion
Previous QPY versions relied on either ``sympy`` or ``symengine`` to serialize the underlying symbolic
expression. Starting in Version 13, QPY now represents the sequence of API calls used to create the
:class:`.ParameterExpression`.

The main change in the serialization format is in the :ref:`qpy_param_expr_v3` payload.  The ``expr_size`` bytes following
the head now contain an array of ``PARAM_EXPR_ELEM_V13`` structs. The intent is for this
array to be read one struct at a time, where each struct describes one of the calls to make to reconstruct the :class:`.ParameterExpression`.
```

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -322,6 +342,139 @@
+    } PARAM_EXPR_ELEM_V13;
+
+The ``op_code`` field is used to define the operation used for the operation
+added to the :class:`.ParameterExpression`. The value can be:
```
```suggestion
The ``op_code`` field is used to define the operation added to the :class:`.ParameterExpression`. The value can be:
```

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -322,6 +342,139 @@
+The ``NULL`` value of 255 is only used to fill the op code field for
+entries that are not actual operations but indicate recursive definitions.
+Then the ``lhs_type`` and ``rhs_type`` fields are used to describe
+the operand types and can be one of the following utf8 encoded
```
```suggestion
the operand types and can be one of the following UTF-8 encoded
```

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -322,6 +342,139 @@
+and is always paired with an ``rhs_type`` of ``n``. The data value is the size in bytes of
+a :ref:`qpy_mapping` encoded mapping of :class:`.Parameter` names to their value for the
+:meth:`~.ParameterExpression.subs` call. The mapping data is immediately following the
+struct, and the next struct starts immediately after the mapping data.
```
```suggestion
When ``s`` is encountered, this indicates that until an ``e` struct is reached, the next structs 
are used for a recursive definition. For both
``s`` and ``e`` types, the data values are not used, and always set to 0. The type value
of ``u`` is used to represent a substitution call. This is only used for ``lhs_type``
and is always paired with an ``rhs_type`` of ``n``. The data value is the size in bytes of
a :ref:`qpy_mapping` encoded mapping of :class:`.Parameter` names to their value for the
:meth:`~.ParameterExpression.subs` call. The mapping data is immediately following the
struct, and the next struct starts immediately after the mapping data.
```

ElePT commented on qiskit/qpy/binary_io/value.py:
```diff
@@ -50,20 +56,132 @@ def _write_parameter_vec(file_obj, obj):
-def _write_parameter_expression(file_obj, obj, use_symengine, *, version):
-    if use_symengine:
-        expr_bytes = obj._symbol_expr.__reduce__()[1][0]
+def _encode_replay_entry(inst, file_obj, version, side=False):
```
If `side` is used to mark the right-side operand, I think it would be clearer to say it explicitly, maybe:

```suggestion
def _encode_replay_entry(inst, file_obj, version, r_side=False):
```
(again, minor suggestion, it just took me a bit to go and check which side it was). If I confused the side, the same would apply to the left.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13356/commits/11f65adbc08690ffe3cb3587814ec1274905b111

ElePT commented on qiskit/qpy/binary_io/value.py:
```diff
@@ -50,20 +56,132 @@ def _write_parameter_vec(file_obj, obj):
+            expr_bytes = buf.getvalue()
+    symbol_table_len = len(obj._parameter_symbols)
+    if extra_symbols:
+        symbol_table_len += 2 * len(extra_symbols)
```
If I'm not mistaken, the `extra_symbols` represent the mapping data, right? why is the len multiplied by 2?

ElePT replied:
Aha, I think I got it, you are iterating over `extra_symbols` twice, one time to encode the keys, and another for the values.

mtreinish replied:
Yep, that was why I set the length to 2x here

ElePT commented on qiskit/qpy/binary_io/value.py:
```diff
@@ -334,6 +495,138 @@ def _read_parameter_expression_v3(file_obj, vectors, use_symengine):
+        return _read_parameter_expr_v13(buf, symbol_map, version, vectors)
+
+
+def _read_parameter_expr_v13(buf, symbol_map, version, vectors):
```
I see why this is the case (there is no previous instance of `_read_parameter_expr`), but I find it a bit difficult to track that `_read_parameter_expr_v13` is related to `_read_parameter_expression_v4`. I don't think I have a good suggestion, because randomly assigning it a v4 would also not be ideal, so I am ok with leaving it as-is.

mtreinish replied:
Oh, I meant to change the v4's to v13s. v4 was just a misreading of the code when I first started this. I thought we were versioning the parameter expression independently because v3 was so low. But I realized it was really changed last in qpy v3. I'll update this.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit/pull/13356/commits/6e75ea58fa69df4c17e5add40b86c0ac5d53c899

ElePT commented on qiskit/qpy/binary_io/value.py:
```diff
@@ -334,6 +495,138 @@ def _read_parameter_expression_v3(file_obj, vectors, use_symengine):
+                    method_str = "__rmul__"
+                elif expression_data.OP_CODE == 3:
+                    method_str = "__rtruediv__"
+                stack.append(getattr(rhs, method_str)(lhs))
```
why is `__rpow__` not included in these? I just noticed that we are entering the `if` for codes 4, 13, 15 too, which correspond to pow, deriv and conj, but then there are no method str changes for them.

mtreinish replied:
For 13 and 15 we can never have a reversed version. They only work if the left hand side is a parameter expression since we can't do the gradient of substitution on an a literal type. I think you're right the lack of `__rpow__` here for `OP_CODE == 5` is probably a bug, I'll add a test case for it and fix it if so.

ElePT replied:
That's what I thought, thanks for confirming the 13 and 15 ones, I wasn't 100% sure

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit/pull/13356/commits/8c2205faf7971fe281d4b82100ab8f79c5bb8d70 this turned out to be a larger bug around non-commutative operations which that commit fixes. Thanks for catching this.

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -322,6 +342,139 @@
+
+if the type value is ``f`` or ``c`` the corresponding ``lhs`` or `rhs``
+field is the literal value with 0 padding for the full 128 bit width in case of a
+float. For ``i`` it's encoded as a 64 bit signed integer with 0 padding for the full
```
I found the wording a bit confusing, I don't know if this is too verbose but I tried to reword this part a little up to my understanding:

```suggestion
If the type value is ``f`` ,``c`` or ``i``, the corresponding ``lhs`` or `rhs``
field widths are 128 bits each. In the case of floats, the literal value is encoded as a double 
with 0 padding, while complex numbers are encoded as real part followed by imaginary part,
taking up 64 bits each. For ``i`, the value is encoded as a 64 bit signed integer with 0 padding for the full
```

ElePT commented on qiskit/qpy/__init__.py:
```diff
@@ -322,6 +342,139 @@
+:class:`.Parameter` which can be looked up in the symbol map described in the
+``map_elements`` outer :ref:`qpy_param_expr_v3` payload. If the type value is
+``s`` this marks the start of a a new recursive section for a nested
+:class:`.ParameterExpression`. For example::
```
```suggestion
:class:`.ParameterExpression`. For example, in the following snippet there is an inner ``expr`` 
contained in ``final_expr``, constituting a nested expression::
```

ElePT reviewed (commented):
Thanks @mtreinish. After taking a look I think the PR looks good. I think we agree in the need for a Qiskit-native alternative for the symengine serialization even if there are some tradeoffs in terms of memory footprint. I added mostly documentation-related suggestions, some typos and some rewordings, and a few minor comments. The question I left in `read_parameter_expression_v4` might be important, but I might have missed something there. Other than that, I was wondering if we should also communicate somehow that `use_symengine`  isn't useful anymore for QPY versions over 13?

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

mtreinish committed:
Change DERIV to GRAD

mtreinish committed:
Change side kwarg to r_side

mtreinish committed:
Change all the v4s to v13s

mtreinish committed:
Correctly handle non-commutative operations

This commit fixes a bug with handling the operand order of subtraction,
division, and exponentiation. These operations are not commutative but
the qpy deserialization code was treating them as such. So in cases
where the argument order was reversed qpy was trying to flip the
operands around for code simplicity and this would result in incorrect
behavior. This commit fixes this by adding explicit op codes for the
reversed sub, div, and pow and preserving the operand order correctly
in these cases.

mtreinish committed:
Fix lint

mtreinish committed:
Merge branch 'main' into qpy-paramexpr-native

mtreinish commented:
> Other than that, I was wondering if we should also communicate somehow that use_symengine isn't useful anymore for QPY versions over 13?

I added a sentence to the `qpy.dump` function's docstring for the `use_symengine` argument to say it doesn't do anything in qpy 13. Were you thinking we needed an upgrade release note or something to documente it further? 

We can drop the argument in 2.0 potentially since 2.0 will have a minimum compatibility version of qpy format version 13 so nothing >=2.0 emits can support the `use_symengine` flag. But my worry with that is it is a potentially breaking api change for users that might want to support dumping qpy across code run with multiple qiskit versions. So maybe it's better to just document it as not doing anything in 2.0 and remove it in 3.0. Either way that's something we can debate post 1.3, since we probably should not deprecate it here, and only save that for a future release.

ElePT reviewed (approved):
> I added a sentence to the `qpy.dump` function's docstring for the `use_symengine` argument to say it doesn't do anything in qpy 13. Were you thinking we needed an upgrade release note or something to documente it further?

I didn't have a fixed idea, but I think the docstring is a good place for the moment. I also wasn't pushing for a deprecation now, just wanted to raise the point to make sure we didn't forget to deal with it in the future.

Thanks a lot for applying the review comments so fast! I have taken a second look and cannot come up with any other potential issue, so on my side I think the PR is ready to be merged.
